### ISSUE TITLE
Support restricting DropZones by tag or component

### ISSUE THREAD

**Issue by chrisvxd** [measuredco/puck#103]
A user should be able to restrict the components that can be dropped into a DropZone.

For example, the user may wish to restrict the root DropZone to layout components, and nested DropZones to leaf components.

Follow on from #87

### Option 1 - single `restrict`  prop

```tsx
// only allow ButtonGroup
<DropZone restrict={["ButtonGroup"]} />
```

* `restrict` might be unclear, so we might be better off naming this prop `allow`

### Option 2 - both `allow` & `disallow`

```tsx
// only allow ButtonGroup
<DropZone allow={["ButtonGroup"]} />

// allow everything except for Columns
<DropZone disallow={["Columns"]} />
```

* Restrict to either `allow` _or_ `disallow` but never both, because that creates strange boolean logic

## Considerations

* To be able to restrict the root DropZone, the user will need to render DropZone instead of `children` in their root `render` method. We will need to establish how they will define the `zone` in this scenario. Perhaps we should export it?

**Comment by xaviemirmon** [comment]
@chrisvxd I love this proposal! This is the next big thing for Puck following nested dropzones!  Option 3 gets my vote and I also think there needs to be a _not_ selector (if possible).  I can see this wanting to have multiple tags here, too which would work with my proposal for #87.

Having something like this would be really great!

`<DropZone zone="content" tags=["layout", "!grid", "!flex"] />`

**Comment by chrisvxd** [comment]
@JakeSidSmith I've updated to the latest 2 proposals I'd like to consider for this following #87. The scope of the discussion is much narrower now, but I'm still keen to hear your thoughts on the remaining 2 options.

Regarding categories - don't think we should add any specific category restrictions, and go with component restrictions only. The user can then restrict categories using this API:

```tsx
<DropZone allow={config.categories.layout.components} />
```

Let me know if you're happy to rework #141, otherwise I'll look into it 🙏

**Comment by JakeSidSmith** [comment]
No disallow?
Also, are we sticking with `allow` as the prop? This might cause complications in the future if we later want to allow/disallow more than components.

**Comment by chrisvxd** [comment]
@JakeSidSmith I left some thoughts in #103. Let's pick up the conversation there 👍

**Comment by chrisvxd** [comment]
@JakeSidSmith we could add `disallow` but (per proposal above) I wasn't sure how to handle a case when a user specifies both `allow` and `disallow`.

What use case are you seeing in future beyond components?

**Comment by JakeSidSmith** [comment]
If a component is specified in both allow and disallow then it is not allowed as both requirements are not met.

There are a couple of benefits to the disallow option:
- you may only want to disallow dropping the same component inside itself
- you may want to allow a collection of components programmatically, but disallow one of them e.g. `allow={Object.values(something)} disallow={['SpecificComponent']}`

Although we don't want to currently support restricting by category I'm sure that users will request this, and it'd be really useful to restrict components by tag name e.g. `allowTagNames={['li']}` for list components (as opposed to having to manually update all the possible options).

In any case, using `allow` on it's own may have a knock on effect to future restrictions, so best to give it a more specific name. No harm in having a longer prop name - if anything it makes it easier to understand at a glance.

**Comment by chrisvxd** [comment]
This has been merged and is now available in `0.13.0-canary.b95050d`!
